pygpu/collectives.pyx

```diff
@@ -0,0 +1,312 @@
+from libc.stdlib cimport malloc, calloc, free
+
+from pygpu.gpuarray cimport (gpucontext, GpuContext, _GpuArray, GpuArray,
+                             ensure_context,
+                             GA_NO_ERROR, get_exc, gpucontext_error,
+                             GpuArray_IS_C_CONTIGUOUS,
+                             GA_C_ORDER, GA_F_ORDER, GA_ANY_ORDER,
+                             pygpu_empty_like, pygpu_empty, memcpy)
+from pygpu.gpuarray import GpuArrayException
+
+
+cdef class GpuCommCliqueId:
+    """
+    """
+    def __cinit__(self, GpuContext context=None, bytes bytearr=None):
+        self.context = ensure_context(context)
+        if bytearr is not None:
+            self.set_comm_id(bytearr)
+        else:
+            comm_generate_id(self.context.ctx, self)
+            self.comm_id_py = self.comm_id.internal[:GA_COMM_ID_BYTES]
+
+    def __eq__(self, other):
+        return type(self) == type(other) and self.comm_id_py == other.comm_id_py
+
+    def __hash__(self):
+        return hash(self.__class__.__name__) ^ hash(self.comm_id_py)
+
+    def __reduce__(self):
+        raise RuntimeError, "Cannot pickle %s object" % self.__class__.__name__
+
+    property comm_id:
+        "Unique clique id to be used by each GpuComm in a group of devices"
+        def __get__(self):
+            return self.comm_id_py
+
+        def __set__(self, bytes bytearr):
+            cdef int length
+            length = len(bytearr)
+            if length < GA_COMM_ID_BYTES:
+                raise ValueError, "GpuComm clique id must have length %d bytes" % (GA_COMM_ID_BYTES)
+            self.comm_id_py = bytearr
+            self.comm_id.internal = <char*>self.comm_id_py
```

Comment:
I am not sure, I will check generated code and ensure if it works with tests.
I intend to write code in a way that it will not copy the `bytes bytearr` into a new array, but it will keep a reference and make `self.comm_id.internal` point correctly.

Also, do you think that making `__get__` and `__set__` refer to the `GpuCommCliqueId` object itself would be a good idea?